main.cpp

```diff
@@ -272,406 +221,213 @@ Use a service like https://www.diffchecker.com/diff to compare your output.
+    {
+        *value = static_cast<Type>( std::pow(*value, ft) );
+        return *this;
+    }
+};
 
-DoubleType& DoubleType::pow(const DoubleType& dt)
+template<>
+struct Numeric<double>
 {
-    return powInternal( static_cast<double>(dt) );
-}
+    using Type = double;
 
-DoubleType& DoubleType::pow(const IntType& it)
-{
-    return powInternal( static_cast<double>(it) );
-}
+    Numeric(Type v) : value( std::make_unique<Type>(v) ){}
 
-DoubleType& DoubleType::apply(std::function<DoubleType&(double&)> func)
-{
-    if(func)
+    ~Numeric()
     {
-        return func(*value);
+        value = nullptr;
     }
 
-    return *this;
-}
-    
-DoubleType& DoubleType::apply(void(*func)(double&))
-{
-    if(func)
+    Numeric& operator+=(Type rhs)
     {
-        func(*value);
+        *value += rhs;
+        return *this;
     }
 
-    return *this;
-}
-
-//------------------------------------------------
-
-IntType& IntType::operator+=(int rhs)
-{
-    *value += rhs;
-    return *this;
-}
-
-IntType& IntType::operator-=(int rhs)
-{
-    *value -= rhs;
-    return *this;
-}
-
-IntType& IntType::operator*=(int rhs)
-{
-    *value *= rhs;
-    return *this;
-}
+    Numeric& operator-=(Type rhs)
+    {
+        *value -= rhs;
+        return *this;
+    }
 
-IntType& IntType::operator/=(int rhs)
-{
-    if (rhs == 0)
+    Numeric& operator*=(Type rhs)
     {
-        std::cout << "error: integer division by zero is an error and will crash the program!" << std::endl;
-    } 
-    else
+        *value *= rhs;
+        return *this;
+    }
+
+    Numeric& operator/=(Type rhs)
     {
+        if (fabs(rhs - 0.0) < std::numeric_limits<Type>::epsilon())
+        { 
+            std::cout << "warning: floating point division by zero!" << std::endl;
+        }
         *value /= rhs;
+        return *this;
     }
 
-    return *this;
-}
+    Numeric& pow(Type ft)
+    {
+        return powInternal( ft );
+    }
 
-IntType& IntType::pow(int it)
-{
-    return powInternal( it );
-}
+    Numeric& powInternal(const Numeric& ft)
+    {
+        *value = std::pow( *value, ft );
+        return *this;
+    }
 
-IntType& IntType::powInternal(int it)
-{
-    if(this->value != nullptr)
-        *value = static_cast<int> (std::pow( static_cast<double>(*value), static_cast<double>(it)));
+    template<typename Callable>
+    Numeric& apply(Callable&& f)
+    {
+        f(value);
+        return *this;
+    }
 
-    return *this;
-}
+    operator Type() const { return *value; }
 
-IntType& IntType::pow(const FloatType& ft)
-{
-    return powInternal( static_cast<int>(ft) );
-}
+private:
+    std::unique_ptr<Type> value = nullptr;
+};
 
-IntType& IntType::pow(const DoubleType& dt)
+//-----------------------------------------------
+template<typename TypeID1, typename TypeID2>
+struct Point
 {
-    return powInternal( static_cast<int>(dt) );
-}
+    Point(const TypeID1& a, const TypeID2& b) : x(static_cast<float>(a)), y(static_cast<float>(b)) {}
```

Comment:
instead of templating the class, just template the constructor function
- [ ] revise

Reply:
What is the benefit of just templating the constructor and not the class in this case? I am not sure what this does. Thank you